contracts/multisigs/GuardCM.sol

```diff
@@ -0,0 +1,292 @@
+
+/// @dev The combination of target and selector is not authorized.
+/// @param target Target address.
+/// @param selector Function selector.
+error NotAuthorized(address target, bytes4 selector);
+
+/// @dev The proposal is not defeated.
+/// @param proposalId Proposal Id.
+/// @param state Current proposal state.
+error NotDefeated(uint256 proposalId, ProposalState state);
+
+/// @title GuardCM - Smart contract for Gnosis Safe community multisig (CM) guard
+/// @author Aleksandr Kuperman - <aleksandr.kuperman@valory.xyz>
+/// @author Andrey Lebedev - <andrey.lebedev@valory.xyz>
+contract GuardCM {
+    event GovernorUpdated(address indexed governor);
+    event SetTargetSelectors(address[] indexed targets, bytes4[] indexed selectors, bool[] statuses);
+    event GovernorCheckProposalIdChanged(uint256 indexed proposalId);
+    event GuardPaused(address indexed account);
+    event GuardUnpaused();
+
+    // Owner address
+    address public immutable owner;
+    // Multisig address
+    address public immutable multisig;
+    // Governor address
+    address public governor;
+    // Guard pausing possibility
+    uint8 public paused = 1;
+    
+    // Mapping of address + bytes4 selector => enabled / disabled
+    mapping(uint256 => bool) public mapAllowedTargetSelectors;
+
+    // Schedule selector
+    bytes4 public constant SCHEDULE = bytes4(keccak256(bytes("schedule(address,uint256,bytes,bytes32,bytes32,uint256)")));
+    // ScheduleBatch selector
+    bytes4 public constant SCHEDULE_BATCH = bytes4(keccak256(bytes("scheduleBatch(address[],uint256[],bytes[],bytes32,bytes32,uint256)")));
+    // Initial check governance proposal Id
+    // Calculated from the proposalHash function of the GovernorOLAS with the following parameters:
+    // targets = [address(0)], values = [0], calldatas = [0x], description = "Is governance alive?"
+    uint256 public governorCheckProposalId = 62151151991217526951504761219057817227643973118811130641152828658327965685127;
+
+    /// @dev GuardCM constructor.
+    /// @param _timelock Timelock address.
+    /// @param _multisig Community multisig address.
+    /// @param _governor Governor address.
+    constructor(address _timelock, address _multisig, address _governor) {
+        owner = _timelock;
+        multisig = _multisig;
+        governor = _governor;
+    }
+
+    /// @dev Changes the governor.
+    /// @param newGovernor Address of a new governor.
+    function changeGovernor(address newGovernor) external {
+        if (msg.sender != owner) {
+            revert OwnerOnly(msg.sender, owner);
+        }
+
+        // Check for the zero address
+        if (newGovernor == address(0)) {
+            revert ZeroAddress();
+        }
+
+        governor = newGovernor;
+        emit GovernorUpdated(newGovernor);
+    }
+
+    /// @dev Changes the governor check proposal Id.
+    /// @param proposalId Governor check proposal Id.
+    function changeGovernorCheckProposalId(uint256 proposalId) external {
```

Comment:
Is this the correct workflow: 
- CM make a proposal on governor (so proposer role is also required) using the parameters suggested above such that the above proposalID is returned
- If the proposal does not succeed the CM have the ability to pause the guard
- Once the guard is paused, CM has fully executor role, so it can schedule and execute via timelock the change of the proposalId

Reply:
why not simply keeping the CM proposer role?

Reply:
The proposal is done via the governor, as we are checking the governor itself. In order to propose in the governor contract, one needs veOLAS

Reply:
yeah right, the proposer role of the CM is for scheduling on the timelock